---
title: "GO:Analysis"
output: html_notebook
---

I'm going to start by making tables of genes of interest - meaning all genes that have appeared as under selection in more than one lineage. I will do this separately for solitary and social lineages before combining at the end. 

GO analysis will be ran on canon / noncanon / random genes under selection per lineage. Go terms can then be checked for comparisons between socialities, etc.

##Genes of Interest

What genes are worth looking at ? Are there genes that pop up in more than one origin / elaboration of sociality, suggesting the change in function may have been selected for in accordance with transitioning to more social lifestyles ? And if so, what are these genes?

This script will take the data, keep only those significant and aggregate by SocOrigin in an object called "undersel". All of those genes that appear more than once will be put in a more interesting dataframe called "goi"

The rest will be put in a dataframe called "undersel"

```{r}
source("Scripts/getgoi.R", local = knitr::knit_global())
```

```{r}
head(goi)
```

Should probably remove the random genes.

```{r}
goi.imm <- goi[!goi$Class == "Random",]
```

```{r}
head(goi.imm)
```

```{r}
goi.imm[goi.imm$Class == "Canon",]
```

```{r}
goi.imm[goi.imm$count > 2,]
```

```{r}
nrow(goi.imm[goi.imm$count > 1,])
nrow(goi.imm[(goi.imm$count > 1) & (goi.imm$Class == "Canon"),])
nrow(goi.imm[(goi.imm$count > 1) & (goi.imm$Class == "NonCanon"),])
```

```{r}
write.table(goi, "output/GOI.tsv", sep = "\t", row.names = F, quote = F)
```

And to do the same for solitary ...

```{r}
source("Scripts/getgoiSol.R", local = knitr::knit_global())
```

```{r}
head(goi.sol)
```

Store the immune genes separately and then write up the solitary GOI.

```{r}
goi.sol.imm <- goi[!goi$Class == "Random",]

write.table(goi, "output/GOI_Sol.tsv", sep = "\t", row.names = F, quote = F)
```


Combine the dataframes and re-aggregate
Remove count and recalculate

```{r}
goi.all <- rbind(goi, goi.sol)

goi.all <- aggregate(goi.all[2], goi.all[-2],
                     FUN = function(X) paste(unique(X), collapse=", "))

goi.all$count <- NULL

goi.all$count <- str_count(goi.all$SocOrigin, pattern = ",")+1

head(goi.all)
```

What's the most often to experience positive selection on immune - related genes? 

```{r}
goi.all[!goi.all$Class == "Random" & goi.all$count > 3,]
```

Write up

```{r}
write.table(goi.all, "output/GOI_All.tsv", sep = "\t", col.names = T, row.names = F, quote = F)
```


##Canon Function

Looking at the canon genes that are under selection (above or equal to one lineage at a time) and assign immune function.

In the previous GOI scripts genes under selection were saved in objects undersel and undersel.sol respectively. 

```{r}
undersel.all <- rbind(undersel, undersel.sol)
undersel.all <- aggregate(undersel.all[2], undersel.all[-2],
                          FUN = function(X) paste(unique(X), collapse = ", "))
undersel.all$count <- NULL
undersel.all$count <- str_count(undersel.all$SocOrigin, pattern = ",")+1
tail(undersel.all)
```

Out of all the genes assessed across all the different lineages, 1950 genes total were found to be showing signs of being under positive selection.

```{r}
undersel.all.can <- undersel.all[undersel.all$Class == "Canon",]
nrow(undersel.all.can)
```

```{r}
head(undersel.all.can)
```


Assign function to the canonical immune genes and save

```{r}
imm.class <- read.table("ImmResources/Canon_Immune_GeneList_Jan20.txt", header = T, sep="\t", quote = "")
imm.fun <- imm.class[,c(1:2)]
head(imm.fun)
```


```{r}
undersel.all.can <- merge(undersel.all.can, imm.fun, by = "Gene")
head(undersel.all.can)
```

```{r}
summary(as.factor(undersel.all.can$Function))
```

```{r}
write.table(undersel.all.can, "output/CanonUnderSel.tsv", row.names = F, quote = F, sep = "\t")
```

Out of curiosity, I wonder how this comparison works looking at the proportion of genes found to be under selection versus how many were considered . . . 

```{r}
func.list <- droplevels(as.factor(imm.fun$Function[!duplicated(imm.fun$Function)]))
func.list
```

```{r}
can.prop <- data.frame(matrix(ncol = 0, nrow = 0))

for (i in 1:length(func.list)){
  func <- func.list[i]
  desc <- paste(func)
  tot <- nrow(imm.fun[imm.fun$Function == paste(func),])
  sel <- nrow(undersel.all.can[undersel.all.can$Function == paste(func),])
  prop <- sel/tot
  comb <- cbind(desc, tot, sel, prop)
  can.prop <- rbind(can.prop, comb)
}

can.prop
```

So the highest proportion of canon immune genes under selection are those involved in the JNK pathway. 

##GO Analysis

The first step here will be making gene lists of genes under selection per each lineage, separated by class of gene (noncanon candidate / random background). There will be a file called Hym_NonCan I will also look at - these are the noncanon genes that are only found in Hymenoptera and not in other insect species considered (see previous project for details). I will also be running a GO analysis of all canon immune genes and all candidate immune genes to see if there is much overlap in terms of gene function. NB: gene lists must be in the form of their associated protein's ID. 

Once the lists are complete, I will write them as text files and save them in "input/GOI/" to be fed into GO analysis scripts (topGO). Results will be written as results files, before I make a master file of all GOterms over represented per lineage per class of gene. 

This can then be used for comparative means - Venn tables, etc.


###Gene Lists

####Canon and Noncanon (all)

These are just lists of all canon and non canon immune genes to look and see if they have much in common.

```{r}
head(imm.verse)
```

Canon

```{r}
can.list <- as.data.frame(imm.verse$ProteinID[imm.verse$Class == "Canon"])

write.table(can.list, "input/GOI/Canon_input.txt", col.names = F, quote = F, row.names = F)
```

... and candidate (noncanon)

```{r}
non.list <- as.data.frame(imm.verse$ProteinID[imm.verse$Class == "NonCanon"])
non.list <- as.data.frame(non.list[!duplicated(non.list$`imm.verse$ProteinID[imm.verse$Class == "NonCanon"]`),])
write.table(non.list, "input/GOI/NonCan_input.txt", col.names = F, quote = F, row.names = F)
```

####Genes under selection

I will craft a script that will go through the undersel object and drag out the genes under selection per lineage per gene class (noncan / random). Each set will be stored in a data.frame object and then written as a .txt file in the "input/GOI" directory. 

First, I'm gonna need an object that has all data - the social and solitary - in one dataframe.

```{r}
data.sol.all <- bind_rows(data.sol)
tail(data.sol.all)
```

```{r}
data.complete <- rbind(data.all, data.sol.all)
head(data.complete)
```

As everything has to be in protein ID format, I'll need a master ID table ready that is isoform reduced.

```{r}
iso.verse <- read.table("Genome_Misc/AmelHAv3.1_filtered.checked.faa.table", header = T, quote = "")
head(iso.verse)
```

Looking at this has highlighted an issue - there's a number of alignments in here which come from non-coding transcripts orginally. At the very least I need to be able to highlight them or else remove them entirely. 

This is more important for the immune stuff so I'll look at that first.

```{r}
imm.nc <- read.table("Genome_Misc/ImmRun_noncoding_list.txt", header = F)
imm.nc
```

```{r}
gene<-strsplit(as.character(imm.nc$V1),".",fixed = T)
imm.nc$Gene<-sapply(gene,"[",1)
imm.nc$V1 <- NULL
imm.nc$Gene <- as.factor(imm.nc$Gene)
nrow(imm.nc)
```

So 13 genes were input. They're all non-canon

```{r}
nrow(undersel.all[undersel.all$Gene %in% imm.nc$Gene,])
undersel.all[undersel.all$Gene %in% imm.nc$Gene,]
```

8 of them were considered showing signs of a longterm shift in selection

```{r}
undersel.all[undersel.all$Gene %in% imm.nc$Gene,]
```

And again for the random genes

```{r}
ran.nc <- read.table("Genome_Misc/Random_noncoding_list.txt", header = F)
gene<-strsplit(as.character(ran.nc$V1),".",fixed = T)
ran.nc$Gene<-sapply(gene,"[",1)
ran.nc$V1 <- NULL
ran.nc$Gene <- as.factor(ran.nc$Gene)
nrow(ran.nc)
```

107 were analysed

```{r}
nrow(undersel.all[undersel$Gene %in% ran.nc$Gene,])
```

71 came back as under selection ? 

I want to remove these but I'll wait for Seth's input.

Anyway, back to making the lists for GO analysis ...

```{r}
source("Scripts/genGeneList2.R", local = knitr::knit_global())
```

###Go Analysis

I had previously produced a GOverse file to use with topGO using the script genGOverse.R. This will be read in using the topGO batch script.


runtopGO.R produces many results files. For each lineage/class combination (and the canon and noncanon input lists) there will be one table written that has all three GO term classes included.

```{r}
#source("Scripts/runtopGO.R"), local = knitr::knit_global()
```

###GO Comparisons

####Canon versus Noncanon (input)

####Apidae Socialities

###Halictidae

