---
title: "Interpretation"
output: html_notebook
---

##Genes of Interest

What genes are worth looking at ? Are there genes that pop up in more than one origin / elaboration of sociality, suggesting the change in function may have been selected for in accordance with transitioning to more social lifestyles ? And if so, what are these genes?

This script will take the data, keep only those significant and aggregate by SocOrigin in an object called "undersel". All of those genes that appear more than once will be put in a more interesting dataframe called "goi"

```{r}
source("Scripts/getgoi.R", local = knitr::knit_global())
```

```{r}
head(goi)
```

Should probably remove the random genes.

```{r}
goi.imm <- goi[!goi$Class == "Random",]
```

```{r}
head(goi.imm)
```

```{r}
goi.imm[goi.imm$Class == "Canon",]
```

```{r}
goi.imm[goi.imm$count > 2,]
```

```{r}
nrow(goi.imm[goi.imm$count > 1,])
nrow(goi.imm[(goi.imm$count > 1) & (goi.imm$Class == "Canon"),])
nrow(goi.imm[(goi.imm$count > 1) & (goi.imm$Class == "NonCanon"),])
```

```{r}
write.table(goi, "output/GOI.tsv", sep = "\t", row.names = F, quote = F)
```

So from here I will have to take a closer look to the genes that appear in more than two lineages tested, and run a GO analysis of the noncanon genes that are under selection and have appeared more than once. It may also be worth to look at AllComplex and AllOrigin too ... 

```{r}
all.sel <- undersel[grep("All", undersel$SocOrigin),]
all.sel.imm <- all.sel[!all.sel$Class == "Random",]
all.sel.imm
```

###All Canon 

Look at all the canon genes and consider their immune function...

Hopefully we should have RNAi / recognition genes in here

```{r}
imm.class <- read.table("ImmResources/Canon_Immune_GeneList_Jan20.txt", header = T, sep="\t", quote = "")
head(imm.class)
```

```{r}
imm.fun <- imm.class[,c(1:2)]
head(imm.fun)
```

```{r}
canon.sel <- undersel[undersel$Class == "Canon",]
canon.sel <- merge(canon.sel, imm.fun, by = "Gene")
summary(as.factor(canon.sel$Function))
head(canon.sel)
```

Make the table a little more useful by aggregating functions...

```{r}
canon.sel <- aggregate(canon.sel[6], canon.sel[-6], 
                     FUN = function(X) paste(unique(X), collapse=", "))
canon.sel <- canon.sel[,c(1:2,4,6)]
canon.sel
```



```{r}
write.table(canon.sel, "output/CanonUnderSel.tsv", row.names = F, quote = F, sep = "\t")
```

##GO Terms Analyses

Separately I've made a GO annotation file from the raw eggNOG file I made previously using the proteome for Amel build AmelHAv3.1.

```{r}
source("Scripts/genGOverse.R", local = knitr::knit_global())
```

I will also need al ist of genes of interest. 

Just to cover all bases the lists I will be looking at include

Non-Canonical immune genes (perhaps with those that do not have orthologs across Insecta ? Depends on whether I can find the raw materials in order to make a list)

All non-canonical immune genes under selection (all, and then divided by origin / elaboration / episodic)

Canonical immune genes under selection (though I suspect this won't be all that interesting)

All random genes under selection (sanity check to compare to other papers that assessed selection on this genome set - also split into all, origin of sociality, elaboration of sociality and episodic)

I'll be running a script that will read in the gene lists, run GO term analysis, and produce dataframes with the most significant terms.

And to run the tests

Results will be generated into output/topGo/

```{r}
source("Scripts/runtopGO.R", local = knitr::knit_global())
```



