---
title: "Project Zero: Solitary Analysis"

output: html_document
---

##Solitary Lineages Analysis

I need to run through the same analysis as I did for the social lineages with the solitary, before finding an appropriate way to compare between the two. 

Some scripts can be run as are, some will need to be edited to work on with solitary input (input/solitary/[immunity/random])

###Setting the Verse

In order to annotate the data with class of gene, description of gene (as it is described in Apis mellifera), etc. we first need a "dataverse" of Amel_HAv3.1.

verseSet.R reads in the relevant information from the Genome_Misc and ImmResources folders.

```{r}
source("Scripts/verseSet.R", local = knitr::knit_global())
```


GeneVerse: created by reading in a file with all genes from Amel HAv3.1 complete with gene product description and transcript ID.

```{r}
head(gene.verse)
```

ImmVerse: created by reading in a file with all immune - associated genes used as the input for the ortholog finding program which created the codon alignments used in the codeML branch-site selection analyses, complete with classifications, transcript, peptide and gene IDs.

```{r}
head(imm.verse)
```

###Immunity Dataset: Adding annotations

dataRead_immunitySol.R will read in the results file, collate them into one source and annotate each gene with descriptions, class of immunity (canon / non-canon) and social origin/elaboration.

```{r}
source("Scripts/dataRead_immunitySol.R", local = knitr::knit_global())
```

```{r}
head(data.sol[[2]])
```

####Tests of significance

SigTestSol.R uses the LRT scores and a Chi square distribution (one-tailed, one degree of freedom) to produce pvalues. Pvalues below 0.05 are considered significant (before correction for multiple testing).

To correct for multiple testing, I'll be using the Benjamini-Hochburg procedure.


```{r}
source("Scripts/SigTestsSol.R", local = knitr::knit_global())
```

```{r}
head(data.sol[[3]])
```

How many are still considered significant ? 

```{r}
data.sol.imm <- bind_rows(data.sol)
data.sol.imm$SocOrigin <- as.factor(data.sol.imm$SocOrigin)
data.sol.sig.imm <- filter(data.sol.imm, adj_pvalue < 0.05)
summary(data.sol.sig.imm$SocOrigin)
```

####Write Up

```{r}
write.table(data.sol.imm, "output/CodeML_Results_Sol_AllImmune.tsv", quote = F, row.names = F, sep = "\t")
write.table(data.sol.sig.imm, "output/CodeML_SigResults_Sol_AllImmune.tsv", quote = F, row.names = F, sep = "\t")
```

###Random Dataset

I'm going to repeat the above, but all at once (using a script, dataReadandTest_randomSol.R).

"Data.sol"" will become a list of the random solitary codeML runs, with it also being saved separated in data.sol.ran.

```{r}
source("Scripts/dataReadandTest_randomSol.R", local = knitr::knit_global())
```

```{r}
head(data.sol[[4]])
```

#### Compare Immunity and Random

First off I'm going to combine the random and immune data and put them back into a list of dataframes.

```{r}
data.sol <- rbind(data.sol.imm, data.sol.ran)
write.table(data.sol, "output/CodeML_BranchSiteAnalysis_Sol_CompleteResults.tsv", sep="\t", row.names=F, quote=F)

head(data.sol)

data.sol <- split(data.sol, f = data.sol$SocOrigin)
head(data.sol[[2]])
```

####Fisher's Exact Test

I will be running a series of Fisher's exact tests (2x2 contingency tables) per social origin: 1 imm versus random, 2 can versus random, 3 noncan versus random, 4 can versus noncan.

Each origin will have its own contingency table per test consisting of a [1,1][1,2],[2,1][2,2] matrix. Colnames are outcomes Selection and No Selection (i.e, significant or not), and the groups are as defined above, in the order defined above. 

For example, CorbSoc test of Can versus Random test will have data input of:

[1,1] = no of can under selection

[2,1] = no of can with no selection (total can - above)

[2,1] = no of random under selection

[2,2] = no of random with no selection (total random - above)

I'll be running a script to achieve the above looking at unadjusted pvalues that are considered significant and spit it out in an object called p.sol.results.

```{r}
source("Scripts/pFisherSol.R", local = knitr::knit_global())
```

```{r}
p.sol.results
```

Sig diff between Can and random and Can and NonCan in Habro but no hits otherwise. 

```{r}
habro <- data.sol[[2]]
habro.sig <- habro[habro$pvalue < 0.051,]
summary(as.factor(habro.sig$Class))
```


And for the adjusted pvalues ...

```{r}
source("Scripts/adjFisherSol.R", local = knitr::knit_global())
```

```{r}
adj.sol.results
```

There's a difference between Immune and Random in AllSol, Can and Random in AllSol and in Habro. Interesting. Still lots of test to go, this one isn't apt after all. . . 

I will have to double check if the Fisher's test is still appropriate now that the size of the random dataset is so much bigger than it was originally.

-- It's not. Sample size is too large. Pearson's Chisq is probably  more apt.

####Pearson's ChiSq with Simulated Pvalues

Using the same contigency tables used above, I will be running ChiSq tests of significance. As in some of these cases the sample frequencies are heavily imbalanced (i.e 5 canonical immune genes under selection versus ~6000 random genes not under selection), I will be simulating the pvalue by Monte Carlo simulation (x10000). 

adjChi.R will run this test and results will appear in the environment in chi.adj.results.

```{r}
source("Scripts/adjChiSol.R", local = knitr::knit_global())
```

```{r}
chi.sol.adj.results
```

Hmm. Imm versus random in AllSol is sig, and Can versus random is significant in AllSol and Habro still. 

####Bootstrapping Approach

Bootstrapping is a resampling technique to estimate statistics on a population by sampling a dataset with replacement.

See my original notes for more.

The script has been written so that 100,000 iterations x social origin x class of gene (canon / noncanon) iterations will take place, with results being stored in boot.results. Off to make a cup of tea.

```{r}
source("Scripts/bootstrapSol.R", local = knitr::knit_global())
```

```{r}
boot.sol.results
```



```{r}
write.table(boot.sol.results, "output/Sol_BootStrapResults.tsv", row.names = F, quote = F, sep = "\t")
```

PropRealDataResult is sig in AllSol and Habro for the Canon .... I'm tempted to run again with more iterations to see if this makes it significant via the CI. If I do this I have to do it for the social lineages too so will think about at the end of running these.

So ChiSq found significance in 3 relationships. So lets have a little look.

AllSol: Imm V Random
How many immune genes considered in total

```{r}
allsol <- data.sol[[1]]
totimm <- as.integer(summary(allsol$Class == "Canon")[3]) + as.integer(summary(allsol$Class == "NonCanon")[3])
totimm
```

What proportion were under selection?

```{r}
allsol.sig <- allsol[allsol$adj_pvalue < 0.051,]
selimm <- as.integer(summary(allsol.sig$Class == "Canon")[3]) + as.integer(summary(allsol.sig$Class == "NonCanon")[3])
allsol.imm.prop <- selimm / totimm
allsol.imm.prop
```

Repeat for random

```{r}
totran <- as.integer(summary(allsol$Class == "Random")[3])
selran <- as.integer(summary(allsol.sig$Class == "Random")[3])
allsol.ran.prop <- selran / totran
allsol.ran.prop
```

Oh ok, more immune genes are under selection compared to background across all solitary lineages.

```{r}
AllSolImmVRan <- c(allsol.imm.prop, allsol.ran.prop)
AllSolImmVRan
```


AllSol: Can V Ran

```{r}
totcan <- as.integer(summary(allsol$Class == "Canon")[3])
selcan <- as.integer(summary(allsol.sig$Class == "Canon")[3])
allsol.can.prop <- selcan / totcan
AllSolCanVRan <- c(allsol.can.prop, allsol.ran.prop)
AllSolCanVRan
```

Habro: Can V Ran

```{r}
habro <- data.sol[[2]]
habro.sig <- habro[habro$adj_pvalue < 0.051,]
totcan <- as.integer(summary(habro$Class == "Canon")[3])
selcan <- as.integer(summary(habro.sig$Class == "Random")[3])
totran <- as.integer(summary(habro$Class == "Random")[3])
selran <- as.integer(summary(habro.sig$Class == "Random")[3])
habro.can.prop <- selcan / totcan 
habro.ran.prop <- selran / totran
HabroCanVRan <- c(habro.can.prop, habro.ran.prop)
HabroCanVRan
```


